lib/aiken/rationals.ak

```diff
@@ -0,0 +1,628 @@
+//      finish compare tests
+
+use aiken/math
+use aiken/option
+
+/// Useful for using Rationals in datums/redeemers which cannot have opaque types.
+/// 
+/// Use **check_rational** or **unsafe_check_rational** to convert to a Rational.
+pub type UncheckedRational {
+  numerator: Int,
+  denominator: Int,
+}
+
+/// Opaque type used to ensure the sign of the Rational is managed strictly in the numerator.
+pub opaque type Rational {
+  numerator: Int,
+  denominator: Int,
+}
+
+/// Converts an UncheckedRational to a Rational
+///
+/// ```aiken
+/// UncheckedRational { numerator: 2, denominator: 3 } |> check_rational() == Some(Rational { numerator: 2, denominator: 3 })
+/// ```
+pub fn check_rational(num: UncheckedRational) -> Option<Rational> {
+  let UncheckedRational { numerator, denominator } = num
+  ratio(numerator, denominator)
+}
+
+/// Unsafely converts an UncheckedRational to a Rational
+///
+/// ```aiken
+/// unsafe_check_rational(UncheckedRational { numerator: 2, denominator: 3 }) == Rational { numerator: 2, denominator: 3 }
+/// ```
+pub fn unsafe_check_rational(num: UncheckedRational) -> Rational {
+  let UncheckedRational { numerator, denominator } = num
+  unsafe_ratio(numerator, denominator)
+}
+
+/// Multiplication
+/// product of two rational values
+/// ```aiken
+/// mul(Rational { numerator: x, denominator: y }, Rational { numerator: a, denominator: b }) == Rational { numerator: x*a, denominator: y*b }
+/// ```
+pub fn mul(a: Rational, b: Rational) -> Rational {
+  let Rational { numerator: a_n, denominator: a_d } = a
+  let Rational { numerator: b_n, denominator: b_d } = b
+  Rational { numerator: a_n * b_n, denominator: a_d * b_d }
+}
+
+test mul_1() {
+  ( unsafe_ratio(2, 3) |> mul(unsafe_ratio(3, 4)) ) == Rational {
+    numerator: 6,
+    denominator: 12,
+  }
+}
+
+test mul_2() {
+  ( unsafe_ratio(-2, 3) |> mul(unsafe_ratio(-3, 4)) ) == Rational {
+    numerator: 6,
+    denominator: 12,
+  }
+}
+
+test mul_3() {
+  (
+    unsafe_ratio(2, 5)
+      |> mul(unsafe_ratio(1, 8))
+      |> mul(unsafe_ratio(3, 10))
+      |> mul(unsafe_ratio(21, 100))
+      |> mul(unsafe_ratio(3, 5))
+      |> mul(unsafe_ratio(2, 8))
+      |> mul(unsafe_ratio(4, 10))
+      |> mul(unsafe_ratio(22, 100))
+      |> reduce()
+  ) == Rational { numerator: 2079, denominator: 50000000 }
+}
+
+/// Division
+/// quotient of two rational values
+/// none if the second value is 0
+/// ```aiken
+/// div(Rational { numerator: x, denominator: y }, Rational { numerator: a, denominator: b }) == Rational { numerator: x*b, denominator: y*a }
+/// ```
+pub fn div(a: Rational, b: Rational) -> Option<Rational> {
+  recip(b) |> option.map(fn(x) { x |> mul(a) })
+}
+
+/// Unsafe division
+/// quotient of two rational values
+/// panics if the second value is 0 (unsafe)
+/// ```aiken
+/// unsafe_ratio(x, y) |> unsafe_div(unsafe_ratio(a, b)) == Rational { numerator: x*b, denominator: y*a }
+/// ```
+pub fn unsafe_div(a: Rational, b: Rational) -> Rational {
```

Comment:
```suggestion
fn unsafe_div(a: Rational, b: Rational) -> Rational {
```

Reply:
Maybe instead of unsafe we can say `expect_div` or `div_or_fail`

Reply:
If invariants are enforced then functions are not unsafe 🙃. A function that crash at runtime or a function that return an Option are "equivalent" and safe. The choice is on whether you expect users to want to handle the error or not. 

(Note that I haven't yet reviewed the implementation logic and only focused on the outer API)

Reply:
Apologies for the confusion due to the names. To me, unsafe makes sense because the function has side effects (crashes the script). I am guessing that is why PlutusTx named those functions that way, too. However, I can see that may not be idiomatic for aiken.

Reply:
Usually, and even in Haskell, when something is called "unsafe" isn't because it may crash but it's because it assumes that the caller is taking upon himself/herself to enforce invariants that could lead to an immediate crash or worse, could lead to data corruption down the line. 

There are many functions in Haskell that are actually partial (eg head on list) and that's "fine", they aren't unsafe but simply push the preconditions to the caller. I am not fond of the approach of impure but safe functions as they're hard to work with and require one to be aware of the invariant. The advantage of having an expressive type system means that we can leverage it to make those conditions when possible (albeit with reason).

In the case of Aiken, I don't really buy the argument for having impure functions as the primary interface and the only argument that was made for them so far is that it's slightly less boilerplate as you can't just write one liners and now have to explicitly call "expect" at call site. Honestly, I don't think this is worth the hassle of making the API harder to decipher and I'd rather consider adding new syntax to the compiler than introduce awkward functions for the sake of removing one line of boilerplate.